### PR TITLE
Update pid.c #5 / harakiri

### PR DESCRIPTION
**Pull request by Crashpilot1000**
Didn't test it. Harakiri Pid probably repaired.

### CODE REVIEW

**Comment by Crashpilot1000** [comment]
Just seeing this right now gyroDataQuant[2] or gyroDataQuant[axis] is not needed gyroDataQuant is sufficient.

**Comment by MJ666** [comment]
Looks to be we worked in parallel and have 2 variants.

**Comment by Crashpilot1000** [comment]
@Mj666: Hehe, you are right! You can cut down one variable (gyroDataQuant) and most probably a plethora of the "= 0.0f" stuff. Tried to compile and got an error that a file wasn't found but I will not "path" around in windows right now. So I think that pid #5 issue is nailed now.

**Comment by hydra** [comment]
Which PR should I merge?  This one or #526 ?

**Comment by hydra** [comment]
I note that the code style, with regards to brackets and tabs, is not consistent with the cleanflight codebase in this PR and would need to be fixed before this can be merged.

File: src/main/flight/pid.c [hunk 1]
```diff
@@ -547,43 +556,52 @@ rollAndPitchTrims_t *angleTrim, rxConfig_t *rxConfig)
...
 
 #ifdef AUTOTUNE
-            if (shouldAutotune()) {
+            if (shouldAutotune())
+            {
                 error = DEGREES_TO_DECIDEGREES(autotune(rcAliasToAngleIndexMap[axis], &inclination, DECIDEGREES_TO_DEGREES(error)));
             }
 #endif
             PTermACC = error * (float)pidProfile->P8[PIDLEVEL] * 0.008f;
             tmp0flt = (float)pidProfile->D8[PIDLEVEL] * 5.0f;
```
Comment (hydra) [inline src/main/flight/pid.c]: Can `tmp0flt` be better named?

Comment (ledvinap) [inline src/main/flight/pid.c]:
```
{
   float  tmp = (float)pidProfile->D8[PIDLEVEL] * 5.0f;
   PTermACC = constrain(PTermACC, -tmp, +tmp);
}
```
will be even cleaner ...

Comment (MJ666) [inline src/main/flight/pid.c]: #526 is retaining the code style and should be at the same level than this one. tmp0 and tmp0flt variables are used in various locations in the code for different purposes? Not sure what a proper name could be. @ledvinap not sure if the incode declaration of the variables is more efficient if you do this multiple times?

Comment (ledvinap) [inline src/main/flight/pid.c]: @mj666: gcc is clever enough to generate good code. Local block definition is actually easier to optimize, compiler does not have to infer that variable value is unused in rest of code.

**Comment by Crashpilot1000** [comment]
Take mj666 variant since he did the initial port, and flighttested it! tmp0flt is just a timecorrector for a multiwii timebase of 3ms. Maybe "Mwii3msTimescale" would be suitable? The other function is to store the precalculated limit for the constrain (like you already depicted :) ).

File: src/main/flight/pid.c (hunk 1 above)
Comment (MJ666) [inline src/main/flight/pid.c]: I will update #526 accordingly and commit this after flight testing.

**Comment by Crashpilot1000** [comment]
@MJ666: Absolutely outstanding job dude!

**Comment by MJ666** [comment]
@Crashpilot1000 did the first implementation in Harakiri and I liked it their. I have seen limitation on the exiting PID controllers when we experimented with the Nano's and the AlienWii32. Personaly I fly this PID controllers with the Nanohexacopters quite successfully but for the Nanooctocopters we have sill yaw problems and I like the PID 3 (default for the AlienWii32) better. There is an overshoot and we have no success yet to tune it properly. still trying. Will be happy if other pilots like it and have good tuning results.

**Comment by hydra** [comment]
The other PR has been merged in now, thanks everyone!

**Comment by Crashpilot1000** [comment]
@hydra: I don't know if I am supposed to keep writing on a closed issue. I don't want to bother you!
@Mj666: Concerning you octo yaw issues. From what I've seen in my testings is that the "roughed" up gyrodata on that axis help, too much precision doesn't seem to be good for yaw. Just observations no further brainy stuff behind that. The MultiWii 2.3 and the preselected yaw in Harakiri SHOULD be the same but your octa flight experience suggests otherwise. Have you tried higher cycletimes like 3 or 3,5ms? The idea is that the I-term may get less time jittery. Generally speaking the yaw axis it the Achilles' heel of multirotors with more than 3 motors. That's why inflight motorstatistics (just for a session) are very valuable to track down symetrical and/or mechanical problems within a minute. That feature would be very handy especially in conjunction with the gui to nail down troubles on new builds and/or degraded/problematic/low quality motors. These motorstatistics can also be used as a plausibility check for the throttle in failsafe situations.

**Comment by MJ666** [comment]
@Crashpilot1000 over the weekend I had the chance to make some flight testing in the open field with the Nano's. The Nanohexa's (have 3 of them) are flying very well now and I like the Harakiri PID more than the MultiWii 2.3. If I go to Nanoocto's (have 2 of them) I have an overshoot and wake yaw behavior with both PID controllers. MultiWii 2.3 is just behaving a bit better here. Since both copters behave the same I would not think this is related to mechanical problems or motor issues. I already tried quite some tuning without much success. as more yaw I use as more (sometimes unpredictable) overshoot I get. Very easy to loose orientation during flight.

**Comment by Crashpilot1000** [comment]
Damn if harakiri had compatibility to your brushed motors you could run g-tune and see with what P values it comes up with your copter. It really helps. Have you tried tuning yaw with I-term zero and low P? Maybe that could track that issue down on your setup? On the other hand you will see yaw problems within the current multiwii - mixer. The constrain in the LUX pidcontroller https://github.com/cleanflight/cleanflight/blob/master/src/main/flight/pid.c#L205 is IMHO fixing that multiwii mixer stuff in "some" way and could be applied to all pidcontrollers. My changed mixer idea is not compatible to those reverse -ESC - pseudo 3D stuff in its current state. The core mixing is here: https://github.com/Crashpilot1000/TestCode3/blob/master/src/mixer.c#L400 . I had the problem that on throttlebursts the yaw shifted noticably - and that is fixed by that. I know BRM hates that but it works for me...

**Comment by robert-b** [comment]
@ Crashpilot1000 
you may play with a runga kutta integrator for yaw only.

**Pull request closed without merging**